### PR TITLE
DAOS-7049 pool: Update map with svc when available

### PR DESCRIPTION
**Pull request by liw**
During a pool_svc_step_up_cb call, the following debug message was
observed:

    ds_pool_evict_rank() Exclude pool d16fe21a/0 rank 5: rc -2008

This was because the pool_svc_step_up_call yielded before completion,
and an "evict" ULT called ds_pool_evict_rank. Rank 5 wouldn't be
excluded as a result, unless a pool service leadership change occured
later.

Related, synchronous, retried IV updates are made by
pool_svc_step_up_cb. (One of these calls caused the yielding in the case
above.) They may cause deadlocks in which rdb_callbackd, the ULT making
the pool_svc_step_up_cb call, is blocked by one of the IV update due to
an unavailable rank, while any "evict" ULT is blocked due to the pool
service isn't UP.

This patch focuses on the following changes:

  - Use the pool_svc leader reference, which is already available in the
    context, when excluding a rank. This avoids relying on the pool
    service being UP.
      - "Evict" ULTs should take pool_svc leader references, for they
	are leader activities.
  - Use "exclude" instead of "evict" in the rank exclusion code, to be
    consistent with "POOL_EXCLUDE", "DAOS_SELF_HEAL_AUTO_EXCLUDE", "dmg
    exclude", etc.
  - Use a single RDB TX for converting target addresses to IDs and
    making pool map changes. This avoids questions like what happens if
    there are other pool map changes between
    pool_find_all_targets_by_addr and ds_pool_update_internal.
      - This requires ds_pool_update_internal parameter changes. A
	cleanup is made to reorder the parameters.
  - Convert some extern functions to static, for they are never used
    outside of ds_pool.

In the future, we shall consider a few more changes:

  - Move synchronous, retried IV updates outside of pool_svc_step_up_cb;
    only blocking in rdb is allowed.
  - Make direct pool map changes in pool_svc_check_node_status, without
    using ULTs.
  - Replace "evict" ULTs with one queue plus one daemon.

Signed-off-by: Li Wei <wei.g.li@intel.com>

### CODE REVIEW

**Comment by liw** [comment]
Improved the function comment of pool_svc_update_map_internal.

File: src/pool/srv_pool.c [hunk 1]
```diff
@@ -4619,17 +4638,25 @@ ds_pool_update_handler(crt_rpc_t *rpc)
...
+		goto out;
+
 	list.pta_number = in->pti_addr_list.ca_count;
 	list.pta_addrs = in->pti_addr_list.ca_arrays;
-	rc = ds_pool_update(in->pti_op.pi_uuid, opc_get(rpc->cr_opc), &list,
-			    &inval_list_out, &out->pto_op.po_map_version,
-			    &out->pto_op.po_hint, false);
-	if (rc)
-		D_GOTO(out, rc);
+	rc = pool_svc_update_map(svc, opc_get(rpc->cr_opc),
```
Comment (kccain) [inline src/pool/srv_pool.c:4648]: When control plane (dmg pool exclude) occurs, this handler I think should be invoked with opcode POOL_EXCLUDE. And if it is excluding a rank (all targets) like pool_svc_exclude_rank() then in this scenario should true be provided as the exlude_rank argument?

Comment (liw) [reply src/pool/srv_pool.c:4648]: It seems so. @wangdi1, in which cases do we want to also exclude ONLY all targets but NOT the parent rank? I'm wondering why we don't always do "exclude_rank == true".

Comment (wangdi1) [reply src/pool/srv_pool.c:4648]: I think there is such case during hot pull&plug.  @NiuYawei  is that true?

File: src/mgmt/srv_drpc.c [hunk 2]
```diff
@@ -59,10 +59,10 @@ ds_mgmt_drpc_prep_shutdown(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	D_INFO("Received request to prep shutdown %u\n", req->rank);
 
 #ifndef DRPC_TEST
-	ds_pool_disable_evict();
+	ds_pool_disable_exclude();
```
Comment (wangdi1) [inline src/mgmt/srv_drpc.c:62]: This only disable local swim DEAD event callback.  Do you remember why do we need it?

File: src/pool/srv_pool.c [hunk 3]
```diff
@@ -3917,48 +3921,100 @@ replace_failed_replicas(struct pool_svc *svc, struct pool_map *map)
...
+		if (inval_tgt_addrs->pta_number > 0) {
+			/*
+			 * If any invalid ranks/targets were specified here,
+			 * abort the entire request. This will mean the
+			 * operator needs to resubmit the request with
+			 * corrected arguments, which will be easier without
+			 * trying to figure out which arguments were accepted &
+			 * started processing already.
+			 */
+			rc = -DER_NONEXIST;
```
Comment (wangdi1) [inline src/pool/srv_pool.c:4003]: this probably worth a DB_MD message.  btw: will -DER_NONEXIST be ignored by dmg pool cmd ?  i.e. 

dmg pool exclude tgts will return 0 or ENONEXIST if there are some targets which do not exist?

**Review by wangdi1 — APPROVED** [review]

File: src/pool/srv_pool.c (hunk 1 above)
Comment (liw) [reply src/pool/srv_pool.c:4648]: @kccain, there are too many different users of this code path. Could we keep the current exclude_rank == false behavior and file a bug for it?

Comment (liw) [reply src/pool/srv_pool.c:4648]: Filed https://jira.hpdd.intel.com/browse/DAOS-7273. @wangdi1, could you take that ticket, please? It seems to be a regression from the target exclusion changes.

File: src/pool/srv_pool.c (hunk 3 above)
Comment (liw) [reply src/pool/srv_pool.c:4003]: Done.

File: src/pool/srv_pool.c (hunk 1 above)
Comment (kccain) [reply src/pool/srv_pool.c:4648]: Yes I guess keep exclude_rank = false here in this patch. When I read the patch code I guess I wasn't focused on existing master and see that it always specifies false today. Seems the use of target id = -1 an be one component of determining whether to exclude the rank, though perhaps there is a hot plug case to check also.

**Review by kccain — DISMISSED** [review]

**Comment by liw** [comment]
Merged with master; no other changes.

**Review by kccain — APPROVED** [review]

**Comment by liw** [comment]
EC degraded failure: https://jira.hpdd.intel.com/browse/DAOS-7350

**Comment by liw** [comment]
Merged the latest master to see if the EC degrade failure goes away.

**Review by wangdi1 — APPROVED** [review]

**Review by wangdi1 — APPROVED** [review]

**Pull request merged**